refac(console): generalize controls widget

hds opened the pull request:
Each view in tokio-console has a widget up the top that lists the
available controls for that view. There was a common implementation of
this for table based views (tasks, resources, and async_ops) and
separate implementations for the task and resource views. The resource
view included two controls widgets, one for the Resource details at the
top of the view, and another above the table of Async Ops at the bottom
of the view.

This change centralises the logic for the creation of this controls
widget. This change is mostly a precursor to also displaying the
controls in the help view (at which point we can revisit whether the
entire list needs to be shown at the top of the screen).

Controls (an action and the key or keys used to invoke it) are defined
in structs so that their definition can be separated from the display
logic (which includes whether or not UTF-8 is supported).

This allows the problem of the text in the controls widget wrapping in
the middle of a control definition to be fixed. Previously a subset of
the controls would have wrapped like this:

```text
controls: select column (sort) = ←→ or h, l, scroll = ↑↓ or k, j, view details
= ↵, invert sort (highest/lowest) = i,
```

Notice how "view details = ↵," was split across multiple lines. The same
list of controls will now wrap at a full control definition.

```text
controls: select column (sort) = ←→ or h, l, scroll = ↑↓ or k, j,
view details = ↵, invert sort (highest/lowest) = i,
```

Additionally, the list of controls on the Resource view has been
consolidated up the top of the screen.

Universal controls, those that are available in all views, are also
defined centrally. As well as the quit action, using the space bar to
pause has been added to that list. This was previously somewhat of an
undocumented feature.

hds committed:
refac(console): generalize controls widget

Each view in tokio-console has a widget up the top that lists the
available controls for that view. There was a common implementation of
this for table based views (tasks, resources, and async_ops) and
separate implementations for the task and resource views. The resource
view included two controls widgets, one for the Resource details at the
top of the view, and another above the table of Async Ops at the bottom
of the view.

This change centralises the logic for the creation of this controls
widget. This change is mostly a precursor to also displaying the
controls in the help view (at which point we can revisit whether the
entire list needs to be shown at the top of the screen).

Controls (an action and the key or keys used to invoke it) are defined
in structs so that their definition can be separated from the display
logic (which includes whether or not UTF-8 is supported).

This allows the problem of the text in the controls widget wrapping in
the middle of a control definition to be fixed. Previously a subset of
the controls would have wrapped like this:

```text
controls: select column (sort) = ←→ or h, l, scroll = ↑↓ or k, j, view details
= ↵, invert sort (highest/lowest) = i,
```

Notice how "view details = ↵," was split across multiple lines. The same
list of controls will now wrap at a full control definition.

```text
controls: select column (sort) = ←→ or h, l, scroll = ↑↓ or k, j,
view details = ↵, invert sort (highest/lowest) = i,
```

Additionally, the list of controls on the Resource view has been
consolidated up the top of the screen.

Universal controls, those that are available in all views, are also
defined centrally. As well as the quit action, using the space bar to
pause has been added to that list. This was previously somewhat of an
undocumented feature.

hawkw commented on tokio-console/src/view/controls.rs:
```diff
@@ -0,0 +1,141 @@
+        let separator = Span::from(", ");
+
+        let controls_count: usize = spans_controls.len();
+        for (idx, spans) in spans_controls.into_iter().enumerate() {
```
nit, take it or leave it: i might like to see some documentation explaining this wrapping logic?

hds replied:
Done.

hawkw commented on tokio-console/src/view/controls.rs:
```diff
@@ -0,0 +1,141 @@
+pub(crate) struct ControlDisplay {
+    pub(crate) action: &'static str,
+    pub(crate) keys: Vec<KeyDisplay>,
+}
```
take it or leave it: it seems like we could allow these to be statically constructed with a `const fn` constructor if we changed this to
```suggestion
pub(crate) struct ControlDisplay {
    pub(crate) action: &'static str,
    pub(crate) keys: &'static [KeyDisplay],
}
```
since there's no actual need to construct them dynamically?

hawkw commented on tokio-console/src/view/controls.rs:
```diff
@@ -0,0 +1,141 @@
+            ControlDisplay::new_simple("quit", "q"),
+        ]
+    })
+}
```
if we changed the `ControlDisplay` type to own an `&'static [KeyDisplay]` rather than a `Vec`, like I described above, `UNIVERSAL_CONTROLS` could just be a constant, rather than needing to be lazily initialized.

hds replied:
Done. Although in resource.rs I had to keep the `OnceCell` because I'm concatenating 2 arrays which it seems can't be dont in `const`.

If you know a way of doing this, that would be really cool.

hawkw commented on tokio-console/src/view/controls.rs:
```diff
@@ -0,0 +1,141 @@
+                    + spans.width()
+                    + needed_trailing_separator_width
+                    <= area.width as usize
+                {
```
style nit, take it or leave it: i might consider
```suggestion
                let total_width = current_line.width()
                    + separator.width()
                    + spans.width()
                    + needed_trailing_separator_width;
                    
                if total_width <= area.width as usize {
```
but, not a big deal.

hawkw commented on tokio-console/src/view/controls.rs:
```diff
@@ -0,0 +1,141 @@
+        let controls_count: usize = spans_controls.len();
+        for (idx, spans) in spans_controls.into_iter().enumerate() {
+            if idx == 0 || current_line.width() == 0 {
+                current_line.0.extend(spans.0);
+            } else {
```
style nit, take it or leave it: i might prefer
```suggestion
                current_line.0.extend(spans.0);
                continue;
            }
```
here, so that the majority of the code in this loop, which is currently an `else` block, doesn't need to be indented? but, this is up to you.

hds replied:
Done (I didn't apply the suggestion because I had to re-indent everything anyway).

hawkw reviewed (approved):
thanks, this looks good to me! i had some very minor suggestions, but they're not blockers for merging this!

hawkw reviewed (approved):
thanks, this looks good to me! i had some very minor suggestions, but they're not blockers for merging this!

hds committed:
Apply suggestions from code review

Co-authored-by: Eliza Weisman <eliza@buoyant.io>

hds committed:
ControlDisplay keys Vec to array

Make a bunch of things around ControlDisplay and KeyDisplay by using
array slices instead of vectors.

hds committed:
Merge remote-tracking branch 'origin/main' into controls-widget-refactor

hawkw commented on tokio-console/src/view/controls.rs:
```diff
@@ -129,13 +129,21 @@ impl ControlDisplay {
-            ControlDisplay::new_simple("quit", "q"),
-        ]
-    })
+const fn universal_controls() -> &'static [ControlDisplay] {
```
nit, not a big deal: can't this just be 
`const UNIVERSAL_CONTROLS: &'static [ControlDisplay]`?

hds replied:
Yes it can! Changed.

hawkw reviewed (approved):
looks good to me!

hds committed:
changed universal_controls to a const UNIVERSAL_CONTROLS

hds committed:
inline UNIVERSAL_CONTROLS where possible

hds committed:
Fix docs on UNIVERSAL_CONTROLS

hds committed:
Merge branch 'main' into controls-widget-refactor
